### DISCUSSION TITLE
Using D3 in Remix.

### DISCUSSION THREAD

**Discussion by leo6liu** [remix-run/remix#2686] [Q&A]
Hello,

I am getting the ERR_REQUIRE_ESM error when trying to use D3 in my Remix app.

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/rktliu_008/algo-sim/frontend/node_modules/d3-array/src/index.js from /home/rktliu_008/algo-sim/frontend/build/index.js not supported.
```

Based on the [Remix Gotchas](https://remix.run/docs/en/v1/pages/gotchas#importing-esm-packages), I was under the assumption I could simply add the line `serverDependenciesToBundle: ["d3"],` to my remix.config.js's module.exports and then I'd be off the the races.

Unfortunately this has had no effect on the ESM error.

Any ideas about how to get D3 working in Remix?

Thanks for your time!

**Comment by sergiodxa** [comment]
The error seems to be on `d3-array` not `d3` itself, could you try adding that too?

**Reply by leo6liu** [reply]
That is my bad, I was further testing individual d3 sub-libraries and forgot to revert just to D3 when I copied the error for the post.

Here's the exact error when I try to import D3:
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/rktliu_008/algo-sim/frontend/node_modules/d3/src/index.js from /home/rktliu_008/algo-sim/frontend/build/index.js not supported.
```

Here's the import: `import * as d3 from "d3";`

**Reply by divmgl** [reply]
Does it work if you just import the specific thing you need from D3? Maybe D3 is exposing a ton of ESM modules and you have to add them all to `serverDependenciesToBundle` 🤷

**Reply by leo6liu** [reply]
@divmgl Wish I could, but D3 has so many nested modules. I tried going one layer in by adding just `d3-array` to `serverDepenedenciesToBundle`, then import that, but no luck. I didn't dig deeper into the modules under d3-array, but if I had and it worked, I would have easily ended up with 30+ dependencies and imports.

For now I'll stay on a different framework, but will keep watching Remix closely. I quite enjoyed the majority of the framework.

**Reply by shawnmmatthews** [reply]
I got this working by using the specific packages I needed from D3. 

So `npm install d3-format d3-array d3-time-format`

remix.config.js
`serverDependenciesToBundle: ["d3-array", "d3-format", "d3-time-format"],`

component 
```
import { format } from "d3-format";
import { extent } from "d3-array";
import { timeParse, timeFormat } from "d3-time-format";
```

Works perfectly.

**Reply by shawnmmatthews** [reply]
@leo6liu

**Reply by leo6liu** [reply]
![image](https://user-images.githubusercontent.com/46534486/163315711-555cdf22-a47e-432f-b31e-01c69c634ceb.png)

No cigar. Maybe I have messed up my Remix app somewhere. I'll test on a clean Remix project.

**Reply by dmylian** [reply]
Had the same problem, so added all the modules that the remix server log suggested:
``` json
{
	"serverDependenciesToBundle": ["d3", "d3-array", "d3-geo", "internmap", "d3-axis", "d3-brush", "d3-chord", "d3-color", "d3-contour", "d3-delaunay", "delaunator", "d3-dispatch", "d3-drag", "d3-dsv", "d3-ease", "d3-fetch", "d3-force", "d3-format", "d3-hierarchy", "d3-interpolate", "d3-path", "d3-polygon", "d3-quadtree", "d3-random", "d3-scale", "d3-scale-chromatic", "d3-selection", "d3-shape", "d3-time", "d3-time-format", "d3-timer", "d3-transition", "d3-zoom"]
}
```
The list seems overly exhaustive but d3 works

**Reply by kiliman** [reply]
I have a script that will find all the ESM dependencies for you. 

https://github.com/kiliman/rmx-cli#-get-esm-packages

**Reply by MoSattler** [reply]
Adding d3 packages `serverDependenciesToBundle` doesn't seem to work. I still end up with the same error as @leo6liu

**Reply by joshtemple** [reply]
Tip, you can make the list more concise with regular expressions:
```typescript
 serverDependenciesToBundle: [
    "d3",
    /^d3-*/,
    "delaunator",
    "internmap",
  ]

**Comment by emilbryggare** [comment]
@leo6liu this is a long shot, but I recently had to add a library to the serverDependenciesToBundle array.

I received the ERR_REQUIRE_ESM when restarting the server after adding the library to serverDependenciesToBundle. However, if I restarted the server for a second time it started successfully.

So try twice. 🙃

**Reply by leo6liu** [reply]
Was worth a shot! Unfortunately didn't work for me here 😄

**Reply by BryceHayden** [reply]
So I had to do a clean build by deleting the build folder (I also redid the prisma migration folder as I had changes but that shouldn't be needed), but then running `pnpm build` followed by `pnpm dev` worked and I was able to get d3 working. so maybe give that a try @leo6liu

**Comment by binajmen** [comment]
@leo6liu Did you by any chance solved that? I'm having the same issue with `react-dnd`. Different library, but similar issue.

**Comment by Elbarae1921** [comment]
I'm having the same issue a year later. @leo6liu did you eventually find a solution?
